### PR TITLE
Testing Ruby with Rspec: Tic Tac Toe

### PR DESCRIPTION
**Pull request by krys2fa**
# Testing Tic Tac Toe
In this project:
- we added Rspec tests to our Tic Tac Toe game

### CODE REVIEW

File: spec/board_spec.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require_relative '../lib/board.rb'
+
+describe Board do
+  describe '#integer_count' do
+    it 'returns true if return value is an integer or nil' do
+      board = Board.new
+      count = board.integer_count
+      expect(count).to be_a Integer || Nil
+    end
```
Comment (bolah2009) [inline spec/board_spec.rb]: Let's test for both positive and negative cases. 
Don't test two cases in on `expect` block, by two cases I meant `integer or nil`, test for when it will return an `integer` and when it will return `nil` separately, but in the same `describe '#integer_count'` block.


Also let's make the test description more aligned with what we testing, instead of `returns true if return value is an integer or nil` why not use `returns an integer when there are values in the board`, this way we know a bit about what `integer_count` is used for.



> A bit of common sense: check if specs cover positive and negative scenarios ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/commonsense.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/)) Example: there is a spec for correct usage of method \[e.g. `it "creates object if all required attributes are passed"`\] and also there is spec for wrong usage of method \[e.g. `it "does not create object if required attributes are missing"`\].

File: spec/board_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  describe '#update_board' do
+    it 'should return true if board is updated' do
+      curr_board = Board.new
+      player = Player.new('chris', 'O')
+      player_move = 5
+      curr_board.update_board(player_move, player)
+      expect(curr_board.board[4]).to eql('O')
+    end
```
Comment (bolah2009) [inline spec/board_spec.rb]: See the previous comment.

Also, this test is about `'#update_board'` why not test directly what does? 🤔 
e.g. => "it returns the player symbol when a board is updated"

File: spec/player_spec.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+describe Player do
+  describe '.player_turn' do
+    it 'should return true if it is an instance of a the Player class' do
+      player_one = Player.new('sam', 'X')
+      player_two = Player.new('chris', 'O')
+      curr_board = Board.new
+      test_player = Player.player_turn(player_one, player_two, curr_board)
+      expect(test_player).to be_a(Player)
+    end
+  end
```
Comment (bolah2009) [inline spec/player_spec.rb]: See previous comments.

Also, I think this test is redundant when `player_turn` is called, it will always return a `Player` class, why not test which player class it will return instead? 🤔
This way we can test what player will be returned after called `player_turn` 2, 3 or even 4 times? 
Using this approach, we are testing for both negative and positive cases.

File: spec/validators_spec.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
+require_relative '../lib/validators.rb'
+require_relative '../lib/board.rb'
+require_relative '../lib/player.rb'
+
+describe Validators do
+  curr_board = Board.new
+  player = Player.new('sam', 'X')
+  move = 3
```
Comment (bolah2009) [inline spec/validators_spec.rb]: [Optional]

I think we can use `let` helper method to wrap the variables here, this ensures that the one test case does not interfere with another.



> Students should [Use `let` instead](https://github.com/rubocop-hq/rspec-style-guide#let-blocks). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/before_each.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Awesome job with the test team! 👏🏾👏🏾👏🏾

## All public methods are tested. 👍🏽

I would, however, recommend that this methods are tested for both positive and negative cases.
Also, let's make the test description more descriptive. Please, see inline comments below for more details.


#### :recycle:  Make these changes and submit another code review request. 👍🏾

#### Feel free to reach me on slack if you having issues implementing these features. :relaxed: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

File: spec/board_spec.rb (hunk 2 above)
Comment (ebeagusamuel) [reply spec/board_spec.rb]: Hello @bolah2009.
Thanks for the review.
However our update board method doesn't actually return a player symbol but rather updates our board with that player symbol.
We have modified our test to check both for when the board is updated successfully, and when it doesn't.

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello Team ,  You have done a great job with this project. :clap:
#### You can go ahead and merge !!!
